pandas/_libs/tslibs/conversion.pyx

```diff
@@ -642,3 +644,99 @@ cpdef inline datetime localize_pydatetime(datetime dt, tzinfo tz):
     elif isinstance(dt, ABCTimestamp):
         return dt.tz_localize(tz)
     return _localize_pydatetime(dt, tz)
+
+
+cdef tzinfo convert_timezone(
+        tzinfo tz_in,
+        tzinfo tz_out,
+        bint found_naive,
+        bint found_tz,
+        bint utc_convert,
+):
+    """
+    Validate that ``tz_in`` can be converted/localized to ``tz_out``.
+
+    Parameters
+    ----------
+    tz_in : tzinfo
+        Timezone info of element being processed.
+    tz_out : tzinfo
+        Timezone info of output.
+    found_naive : bool
+        Whether a timezone-naive element has been found so far.
+    found_tz : bool
+        Whether a timezone-aware element has been found so far.
+    utc_convert : bool
+        Whether to convert/localize to UTC.
+
+    Returns
+    -------
+    tz_info
+        Timezone info of output.
+
+    Raises
+    ------
+    ValueError
+        If ``tz_in`` can't be converted/localized to ``tz_out``.
+    """
+    if tz_in is not None:
+        if utc_convert:
+            pass
+        elif found_naive:
+            raise ValueError('Tz-aware datetime.datetime '
+                             'cannot be converted to '
+                             'datetime64 unless utc=True')
+        elif tz_out is not None and not tz_compare(tz_out, tz_in):
+            raise ValueError('Tz-aware datetime.datetime '
+                             'cannot be converted to '
+                             'datetime64 unless utc=True')
+        else:
+            tz_out = tz_in
+    else:
+        if found_tz and not utc_convert:
+            raise ValueError('Cannot mix tz-aware with '
+                             'tz-naive values')
+    return tz_out
+
+
+cdef int64_t parse_pydatetime(
+        object val,
+        npy_datetimestruct *dts,
+        bint utc_convert,
+) except? -1:
+    """
+    Convert pydatetime to datetime64.
+
+    Parameters
+    ----------
+    val
+        Element being processed.
+    dts : *npy_datetimestruct
+        Needed to use in pydatetime_to_dt64, which writes to it.
+    utc_convert : bool
+        Whether to convert/localize to UTC.
+
+    Raises
+    ------
+    OutOfBoundsDatetime
+    """
+    cdef:
+        _TSObject _ts
+        int64_t result
+
+    if val.tzinfo is not None:
+        if utc_convert:
+            _ts = convert_datetime_to_tsobject(val, None)
+            _ts.ensure_reso(NPY_FR_ns)
+            result = _ts.value
+        else:
+            _ts = convert_datetime_to_tsobject(val, None)
+            _ts.ensure_reso(NPY_FR_ns)
```

Comment:
Meh I've changed my mind on this. Can you change `ensure_reso` from returning `void` to returning an `int` and have it be -1 on error? This keeps our error checking patterns consistent with CPython standards.

https://docs.python.org/3/c-api/exceptions.html#exception-handling

Void makes things pretty confusing here

Reply:
You probably don't even need the `except? -1` at all if you change the return type. Looks like Cython handles that automatically:

https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html

```
cdef functions that are not extern are implicitly declared with a suitable exception specification for the return type (e.g. except * or @cython.exceptval(check=True) for a void return type, except? -1 or @cython.exceptval(-1, check=True) for an int return type).
```

I just think the current void return type is going to swallow errors and make it unclear where things fail

Reply:
as in,
```python
    cdef int ensure_reso(self, NPY_DATETIMEUNIT creso):
        if self.creso != creso:
            try:
                self.value = convert_reso(self.value, self.creso, creso, False)
            except Exception:
                return -1
        return 0
```
?

Reply:
I think that's just in the in-development version https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#exception-values-and-noexcept

> cdef functions that are not extern now propagate Python exceptions by default, where previously they needed to explicitly be declated with an [exception value](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values) in order for them to do so.

Reply:
I think just 

```python
cdef int ensure_reso(self, NPY_DATETIMEUNIT creso):
  if self.creso != creso:
    self.value = convert_reso(self.value, self.creso, creso, False)

  return self.value
```

If you look at the generated code I would expect that Cython would wrap this loosely with `if ((self.value == -1) & (PyErr_Occurred())` naturally, whereas the void call just does the plain PyErr_Occurredd check

Reply:
ok done, thanks

the line that throws the exception is

https://github.com/pandas-dev/pandas/blob/ebb8b3905bdf96f9dca3e4d76321d63a7fd355d8/pandas/_libs/tslibs/conversion.pyx#L742

for example on
```python
to_datetime([datetime(9999, 1, 1)])
```

Reply:
> I think that's just in the in-development version https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#exception-values-and-noexcept
> 
> > cdef functions that are not extern now propagate Python exceptions by default, where previously they needed to explicitly be declated with an [exception value](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values) in order for them to do so.

Oh great find on this. So you can see how this works locally if you set up a file like this:

```python
cdef int ensure_reso() except? -1:
    return -1;

cdef void ensure_reso_void() except *:
    pass

cdef int int_wrapper() except? -1:
    return ensure_reso()

cdef void void_wrapper() except *:
    ensure_reso_void()
```

run `cythonize -3 <the_file>.pyx` and you can see what it generates for the two functions:

int_wrapper:

```c
static int __pyx_f_4test_int_wrapper(void) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("int_wrapper", 0);

  /* "test.pyx":8
 * 
 * cdef int int_wrapper() except? -1:
 *     return ensure_reso()             # <<<<<<<<<<<<<<
 * 
 * cdef void void_wrapper() except *:
 */
  __pyx_t_1 = __pyx_f_4test_ensure_reso(); if (unlikely(__pyx_t_1 == ((int)-1) && PyErr_Occurred())) __PYX_ERR(0, 8, __pyx_L1_error)
  __pyx_r = __pyx_t_1;
  goto __pyx_L0;

  /* "test.pyx":7
 *     pass
 * 
 * cdef int int_wrapper() except? -1:             # <<<<<<<<<<<<<<
 *     return ensure_reso()
 * 
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_AddTraceback("test.int_wrapper", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}
```

vs void_wrapper:

```c
static void __pyx_f_4test_void_wrapper(void) {
  __Pyx_RefNannyDeclarations
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("void_wrapper", 0);

  /* "test.pyx":11
 * 
 * cdef void void_wrapper() except *:
 *     ensure_reso_void()             # <<<<<<<<<<<<<<
 */
  __pyx_f_4test_ensure_reso_void(); if (unlikely(PyErr_Occurred())) __PYX_ERR(0, 11, __pyx_L1_error)

  /* "test.pyx":10
 *     return ensure_reso()
 * 
 * cdef void void_wrapper() except *:             # <<<<<<<<<<<<<<
 *     ensure_reso_void()
 */

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_AddTraceback("test.void_wrapper", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
}
```

Reply:
So to your point about the cython version I think you need to add `except? -1` to the end of `ensure_reso`. That should point you to the correct location when things fail